### PR TITLE
Remove vm_run and directly inline its contents

### PR DESCRIPTION
**Pull request by galpeter**
Replaced vm_run calls to directly create the vm frame context and do it's initialization.

Added new methods:
* vm_init_frame
* vm_calculate_frame_size

### CODE REVIEW

**Comment by galpeter** [comment]
Memory/perf stats:

Benchmark | Heap (bytes) | Stack (bytes) | Perf (sec) |
----: | ----: | ----: | ----: | 
3d-cube.js | 10672 -> 10672 : 0.000% | 12888 -> 11592 : **+10.056%** | 0.735 -> 0.728 : +1.000% | 
3d-morph.js | 288984 -> 288984 : 0.000% | 1440 -> 1456 : -1.111% | 0.728 -> 0.710 : **+2.496%** | 
3d-raytrace.js | 116424 -> 116424 : 0.000% | 2544 -> 2376 : **+6.604%** | 0.869 -> 0.863 : +0.708% | 
access-binary-trees.js | 27120 -> 27120 : 0.000% | 3104 -> 2864 : **+7.732%** | 0.448 -> 0.445 : +0.844% | 
access-fannkuch.js | 1560 -> 1560 : 0.000% | 1512 -> 1528 : -1.058% | 1.755 -> 1.748 : +0.365% | 
access-nbody.js | 4088 -> 4088 : 0.000% | 1512 -> 1528 : -1.058% | 0.982 -> 0.975 : +0.637% | 
bitops-3bit-bits-in-byte.js | 792 -> 792 : 0.000% | 1472 -> 1488 : -1.087% | 0.474 -> 0.471 : +0.641% | 
bitops-bits-in-byte.js | 712 -> 712 : 0.000% | 1472 -> 1488 : -1.087% | 0.650 -> 0.645 : +0.797% | 
bitops-bitwise-and.js | 536 -> 536 : 0.000% | 1352 -> 1368 : -1.183% | 0.826 -> 0.806 : **+2.459%** | 
bitops-nsieve-bits.js | 60904 -> 60904 : 0.000% | 1472 -> 1488 : -1.087% | 0.988 -> 0.989 : -0.139% | 
controlflow-recursive.js | 1000 -> 1000 : 0.000% | 57552 -> 51432 : **+10.634%** | 0.318 -> 0.310 : **+2.587%** | 
crypto-aes.js | 26736 -> 26736 : 0.000% | 1864 -> 1864 : 0.000% | 0.587 -> 0.584 : +0.480% | 
crypto-md5.js | 70496 -> 70496 : 0.000% | 1680 -> 1560 : **+7.143%** | 0.479 -> 0.479 : -0.093% | 
crypto-sha1.js | 45832 -> 45832 : 0.000% | 1552 -> 1568 : -1.031% | 0.492 -> 0.498 : -1.154% | 
date-format-tofte.js | 8672 -> 8672 : 0.000% | 1872 -> 1800 : **+3.846%** | 0.644 -> 0.643 : +0.239% | 
date-format-xparb.js | 15504 -> 15504 : 0.000% | 2304 -> 2256 : **+2.083%** | 0.492 -> 0.491 : +0.085% | 
math-cordic.js | 1696 -> 1696 : 0.000% | 1472 -> 1488 : -1.087% | 1.072 -> 1.066 : +0.498% | 
math-partial-sums.js | 1256 -> 1256 : 0.000% | 1432 -> 1448 : -1.117% | 0.623 -> 0.625 : -0.358% | 
math-spectral-norm.js | 2920 -> 2920 : 0.000% | 1472 -> 1488 : -1.087% | 0.472 -> 0.467 : **+1.026%** | 
string-base64.js | 89840 -> 89840 : 0.000% | 1488 -> 1504 : -1.075% | 0.890 -> 0.929 : -4.362% | 
string-fasta.js | 3552 -> 3552 : 0.000% | 1456 -> 1472 : -1.099% | 1.034 -> 1.047 : -1.255% | 
Geometric mean: | 0.000% | +1.703% | +0.367% | 

Binary (bytes) | master(b2e54e915b) | patch(4b9fab0615) | Diff |
----: | ----: | ----: | ----: |
size | 136716 | 136716 | 0 bytes |
.rodata | 11850 | 11850 | 0 bytes |
.dynstr | 212 | 212 | 0 bytes |
.rel.plt | 160 | 160 | 0 bytes |
.interp | 25 | 25 | 0 bytes |
.dynsym | 416 | 416 | 0 bytes |
.gnu.hash | 212 | 212 | 0 bytes |
.text | 117304 | 117544 | +240 bytes |
.comment | 85 | 85 | 0 bytes |
.shstrtab | 226 | 226 | 0 bytes |
.data | 8 | 8 | 0 bytes |
.ARM.exidx | 8 | 8 | 0 bytes |
.rel.dyn | 16 | 16 | 0 bytes |
.init | 12 | 12 | 0 bytes |
.got | 96 | 96 | 0 bytes |
.plt | 268 | 268 | 0 bytes |
.note.ABI-tag | 32 | 32 | 0 bytes |
.gnu.version_r | 32 | 32 | 0 bytes |
.bss | 1575096 | 1575096 | 0 bytes |
.fini | 8 | 8 | 0 bytes |
.hash | 180 | 180 | 0 bytes |
.gnu.version | 52 | 52 | 0 bytes |
.fini_array | 4 | 4 | 0 bytes |
.init_array | 4 | 4 | 0 bytes |
.dynamic | 240 | 240 | 0 bytes |
.eh_frame | 4 | 4 | 0 bytes |
.ARM.attributes | 53 | 53 | 0 bytes |

File: jerry-core/ecma/operations/ecma-function-object.c [hunk 1]
```diff
@@ -852,11 +852,16 @@ ecma_op_function_call_simple (ecma_object_t *func_obj_p, /**< Function object */
...
+  ecma_value_t ret_value;
+  {
+    size_t frame_size = vm_calculate_frame_size (bytecode_data_p);
+    JERRY_VLA (uintptr_t, stack, frame_size);
+
+    vm_frame_ctx_t *frame_ctx_p = (vm_frame_ctx_t *) stack;
+    vm_init_frame (frame_ctx_p, bytecode_data_p, local_env_p, this_binding);
+    vm_init_exec (frame_ctx_p, arguments_list_p, arguments_list_len);
+    ret_value = vm_execute (frame_ctx_p);
+  }
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-function-object.c]: This patch adds a lot of cloning. What about creating an always inline function instead?

Comment (galpeter) [reply jerry-core/ecma/operations/ecma-function-object.c]: Ok, I'll do an update. I've just did a bit of measurement and with lto and inline it'll be ok. Without lto there will be a bit of performance degradation.

**Comment by galpeter** [comment]
Closing for now, as the changes does not provide extra improvement at the moment. Will re-open if another solution was found.

**Pull request closed without merging**